src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+    'FALSE'. The virtual address used to switch can be defined as a list
+    with the single element (as if it were a multiplexer) or by using the
+    shorthand which is to define the pin_name attribute as a string.
+    """
+
+    pin_name: Pin = ""
+    map_tree = ("FALSE", "TRUE")
+
+    def multiplex(
+        self, signal_output: Union[Signal, bool], trigger_update: bool = True
+    ) -> None:
+        if signal_output is True:
+            signal = "TRUE"
+        elif signal_output is False:
+            signal = "FALSE"
+        else:
+            signal = signal_output
+        super().multiplex(signal, trigger_update=trigger_update)
+
+    def __init__(
+        self,
+        update_pins: Optional[PinUpdateCallback] = None,
+    ):
+        if not self.pin_list:
+            self.pin_list = [self.pin_name]
+        super().__init__(update_pins)
+
+
+class RelayMatrixMux(VirtualMux):
+    clearing_time = 0.01
+
+    def _calculate_pins(
+        self, old_signal: Signal, new_signal: Signal
+    ) -> tuple[PinSetState, PinSetState]:
+        """
+        Override of _calculate_pins to implement break-before-make switching.
+        """
+        setup = PinSetState(off=self._pin_set, on=frozenset())
+        on_pins = self._signal_map[new_signal]
+        final = PinSetState(off=self._pin_set - on_pins, on=on_pins)
+        return setup, final
+
+
+class AddressHandler:
+    """
+    Controls the IO for a set of pins.
+
+    For output, it is assumed that all the pins under the of a given
+    AddressHandler are updated in one operation.
+
+    This base class doesn't give you much. You need to create a subclass
+    that implement a set_pins() method.
+
+    :param pin_list: Sequence of pins
+    """
+
+    pin_list: Sequence[Pin] = ()
+
+    def set_pins(self, pins: Collection[Pin]) -> None:
+        """
+        Called by the VirtualAddressMap to write out pin changes.
+
+        : param pins: is a collection of pins that should be made active. All other
+        pins defined by the AddressHandler should be cleared.
+        """
+        raise NotImplementedError
+
+    def close(self) -> None:
+        """
+        Optional close method to clean-up resources.
+
+        This will be called automatically by the JigDriver for any
+        address handlers passed into the JigDriver when it as created.
+        """
+        pass
+
+
+class PinValueAddressHandler(AddressHandler):
+    """Maps pins to bit values then combines the bit values for an update"""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._pin_lookup = {
+            pin: bit for pin, bit in zip(self.pin_list, bit_generator())
+        }
+
+    def set_pins(self, pins: Collection[Pin]) -> None:
+        value = sum(self._pin_lookup[pin] for pin in pins)
+        self._update_output(value)
+
+    def _update_output(self, value: int) -> None:
+        # perhaps it's easy to compose by passing the output
```

Comment:
I think I'll just leave this as is for now. Now that I've added a built in relay matrix handler, it will be pretty rare that scripts need to implement their own thing anyway.